src/scipp/core/bin_remapping.py

```diff
@@ -1,114 +1,124 @@
 # @author Simon Heybrock
-from typing import Dict
+import itertools
+import uuid
+from typing import Dict, List
+from math import prod
 from .._scipp import core as _cpp
-from .cpp_classes import DataArray, Variable, Dataset
-from .like import empty_like
+from .cpp_classes import DataArray, Variable
 from .cumulative import cumsum
-from .dataset import irreducible_mask
-from ..typing import VariableLikeType
+from ..typing import Dims, VariableLikeType
+from .variable import index
+from .operations import where
+from .concepts import (concrete_dims, rewrap_reduced_data, irreducible_mask)
 
 
-def _reduced(obj: Dict[str, Variable], dim: str) -> Dict[str, Variable]:
-    return {name: var for name, var in obj.items() if dim not in var.dims}
-
-
-def reduced_coords(da: DataArray, dim: str) -> Dict[str, Variable]:
-    return _reduced(da.coords, dim)
-
-
-def reduced_attrs(da: DataArray, dim: str) -> Dict[str, Variable]:
-    return _reduced(da.attrs, dim)
-
-
-def reduced_masks(da: DataArray, dim: str) -> Dict[str, Variable]:
-    return {name: mask.copy() for name, mask in _reduced(da.masks, dim).items()}
-
-
-def _copy_dict_for_overwrite(mapping: Dict[str, Variable]) -> Dict[str, Variable]:
-    return {name: copy_for_overwrite(var) for name, var in mapping.items()}
-
-
-def copy_for_overwrite(obj: VariableLikeType) -> VariableLikeType:
-    """
-    Copy a Scipp object for overwriting.
-
-    Unlike :py:func:`scipp.empty_like` this does not preserve (and share) coord,
-    mask, and attr values. Instead, those values are not initialized, just like the
-    data values.
-    """
-    if isinstance(obj, Variable):
-        return empty_like(obj)
-    if isinstance(obj, DataArray):
-        return DataArray(copy_for_overwrite(obj.data),
-                         coords=_copy_dict_for_overwrite(obj.coords),
-                         masks=_copy_dict_for_overwrite(obj.masks),
-                         attrs=_copy_dict_for_overwrite(obj.attrs))
-    ds = Dataset(coords=_copy_dict_for_overwrite(obj.coords))
-    for name, da in obj.items():
-        ds[name] = DataArray(copy_for_overwrite(da.data),
-                             masks=_copy_dict_for_overwrite(da.masks),
-                             attrs=_copy_dict_for_overwrite(da.attrs))
-    return ds
-
-
-def hide_masked_and_reduce_meta(da: DataArray, dim: str) -> DataArray:
-    if (mask := irreducible_mask(da.masks, dim)) is not None:
+def hide_masked(da: DataArray, dim: Dims) -> DataArray:
+    if (mask := irreducible_mask(da, dim)) is not None:
         # Avoid using boolean indexing since it would result in (partial) content
         # buffer copy. Instead index just begin/end and reuse content buffer.
         comps = da.bins.constituents
-        select = ~mask
-        data = _cpp._bins_no_validate(
-            data=comps['data'],
-            dim=comps['dim'],
-            begin=comps['begin'][select],
-            end=comps['end'][select],
-        )
+        if mask.ndim == 1:
```

Comment:
Can you explain the distinction between 1d and other masks?

Reply:
Adding a comment.